compiler/luci/import/src/ImportMetadata.cpp

```diff
@@ -0,0 +1,161 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ImportMetadata.h"
+
+#include <vector>
+
+namespace
+{
+
+/**
+ * 'op_table'
+ * - [ id : uint32_t ][ node_num : uint32_t ][ node_ids : node_num * uint32_t ]
+ * - 'node_num' should be larger than zero. (TODO Check it!)
+ */
+std::map<uint32_t, std::set<uint32_t>> metadata_with_op_table(const std::vector<uint8_t> &buffer)
+{
+  std::map<uint32_t, std::set<uint32_t>> node_source_ids_map;
+
+  uint32_t idx = 0;
+  while (idx < buffer.size())
+  {
+    uint32_t id = 0;
+    id += (buffer.at(idx + 0) << 0 * 8);
+    id += (buffer.at(idx + 1) << 1 * 8);
+    id += (buffer.at(idx + 2) << 2 * 8);
+    id += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    uint32_t node_num = 0;
+    node_num += (buffer.at(idx + 0) << 0 * 8);
+    node_num += (buffer.at(idx + 1) << 1 * 8);
+    node_num += (buffer.at(idx + 2) << 2 * 8);
+    node_num += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    std::set<uint32_t> source_ids;
+    for (uint32_t j = 0; j < node_num; ++j)
+    {
+      uint32_t origin = 0;
+      origin += (buffer.at(idx + 4 * j + 0) << 0 * 8);
+      origin += (buffer.at(idx + 4 * j + 1) << 1 * 8);
+      origin += (buffer.at(idx + 4 * j + 2) << 2 * 8);
+      origin += (buffer.at(idx + 4 * j + 3) << 3 * 8);
+
+      source_ids.insert(origin);
+    }
+    idx += (node_num * 4);
+
+    if (node_source_ids_map.insert({id, source_ids}).second == false)
+      throw std::runtime_error("Duplicated origin ID");
+  }
+  assert(idx == buffer.size());
+
+  return node_source_ids_map;
+}
+
+/**
+ * 'source_table'
+ * - [ id : uint32_t ][ length : uint32_t ][ data : 'length' Bytes ]
+ */
+std::map<uint32_t, std::string> metadata_with_source_table(const std::vector<uint8_t> &buffer)
+{
+  std::map<uint32_t, std::string> source_id_name_map;
+
+  uint32_t idx = 0;
+  while (idx < buffer.size())
+  {
+    uint32_t id = 0;
+    id += (buffer.at(idx + 0) << 0 * 8);
+    id += (buffer.at(idx + 1) << 1 * 8);
+    id += (buffer.at(idx + 2) << 2 * 8);
+    id += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    uint32_t length = 0;
+    length += (buffer.at(idx + 0) << 0 * 8);
+    length += (buffer.at(idx + 1) << 1 * 8);
+    length += (buffer.at(idx + 2) << 2 * 8);
+    length += (buffer.at(idx + 3) << 3 * 8);
+    idx += 4;
+
+    std::string origin_name;
+    for (uint32_t j = 0; j < length; ++j)
+      origin_name += (char)buffer.at(idx + j);
+    idx += length;
+
+    if (source_id_name_map.insert({id, origin_name}).second == false)
+      throw std::runtime_error("Duplicated origin ID");
+  }
+  assert(idx == buffer.size());
+
+  return source_id_name_map;
+}
+
+} // namespace
+
+namespace luci
+{
+
+ImportMetadata::ImportMetadata(const luci::CircleReader &reader)
```

Comment:
(Originally written by @seanshpark)

please implement body to some separate method NOT in ctor